docs/rfc/rfc-018-bls-agg-exploration.md

```diff
@@ -0,0 +1,400 @@
+The block proposer would gather all of the votes, complete with signatures,
+as it does now, and produce a set of aggregate signatures from all of the
+individual vote signatures.
+
+Implementing 'vote-time' signature aggregation cannot be achieved without
+also implementing commit-time signature aggregation. This is because such
+signatures cannot be dis-aggregated into their constituent pieces. Therefore,
+in order to implement 'vote-time' signature aggregation, we would need to
+either first implement 'commit-time' signature aggregation, or implement both
+'vote-time' signature aggregation while also updating the block creation and
+verification protocols to allow for aggregated signatures.
+
+### Rogue key attack prevention
+
+Generating an aggregated signature requires guarding against what is called
+a [rogue key attack][bls-ietf-terms]. A rogue key attack is one in which a
+malicious actor can craft an _aggregate_ key that can produce signatures that
+appear to include a signature from a private key that the malicious actor
+does not actually know. In Tendermint terms, this would look like a Validator
+producing a vote signed by both itself and some other validator where the other
+validator did not actually produce the vote itself.
+
+The main mechanisms for preventing this require that each entity prove that it
+can can sign data with just their private key. The options involve either
+ensuring that each entity sign a _different_ message when producing every
+signature _or_ producing a [proof of possession][bls-ietf-pop] when announcing
+their key to the network.
+
+A proof of possession is a message that demonstrates ownership of a private
+key. A simple scheme for proof of possession is one where the entity announcing
+its new public key to the network includes a digital signature over the bytes
+of the public key generated using the associated private key. Everyone receiving
+the public key and associated proof-of-possession can easily verify that the
+signature and be sure the entity owns the private key.
+
+This proof-of-possession scheme suits the Tendermint use case quite well since
+validator keys change infrequently so the associated PoPs would not be onerous
+to produce, verify, and store. Using this scheme allows signature verification
+to proceed more quickly, since all signatures are over identical data and
+can therefore be checked using an aggregated public key instead of one at a
+time, public key by public key.
+
+### Backwards Compatibility
+
+Backwards compatibility is an important consideration for signature verification.
+Specifically, it is important to consider whether chains using current versions
+of IBC would be able to interact with chains adopting BLS.
+
+Because the Block shared by IBC and Tendermint is produced and parsed using
+protobuf, new structures can be added to the Block without breaking the
+ability of legacy users to parse the new structure. Breaking changes between
```

Comment:
They can parse it, but if participating in the network requires meaningful work using the new fields, there's still a potential compatibility hitch. (Not a wire breakage, at least, though)

Reply:
Yes definitely. The scheme I describe here _does_ mean that we could introduce the functionality without simultaneously breaking IBC between old chains if we wanted to go that route. It would be up to peer chains to update to be able to verify peer chains' headers.

Reply:
That makes sense. It does suggest, though, that if _any_ chain that participates in IBC wants to swap to BLS signatures, _all_ of them will have to for IBC to continue working?

Reply:
Yes, this, to a degree is the fundamental difficulty of IBC at the moment. All chains are effectively required to be on the latest version to interop. Any update to one forces an update to peer chains or risks becoming disconnected.

Reply:
> if any chain that participates in IBC wants to swap to BLS signatures, all of them (at least those it interops with) will have to for IBC to continue working

Might be useful to add a statement like this into this section since it might not be apparent to all readers

Reply:
For some more context, I discussed with @AdityaSripal how we might actually support upgrading to BLS clients in ibc-go. At the minimum we would need to add a new function to the `ClientState` interface which would allow conversion of a legacy `ClientState` to the new `ClientState`, along with any migration necessary. The new client state would reference the new client implementation which does the BLS verification

These are relatively straightforward changes, but we would need some advance notice on when these changes will be needed. They ideally would be bundled with a new client implementation (such as the BLS verification changes)

cc @crodriguezvega

Reply:
> These are relatively straightforward changes, but we would need some advance notice on when these changes will be needed.

Of course, we will definitely continue to be very mindful of the disruption this may cause in the context of IBC. Because of the graceful way by which chains would update to BLS, i.e. by swapping out validator keys, I think we should be able to verify block data on either side of the breaking change without serious modification to the light client. The only change would be to the types of commit signatures the client can understand.